Pull request for supervisor review (#1)
AstroViking/ba-thesis: chapters/02-theory.Rmd

# Theory {#mathematicalTheory}

## Exact Kernel Density Estimation

> An exact Kernel Density Estimation can be calculated as

Reviewer: ```suggestion
An exact Kernel Density Estimation is defined as
```

\begin{equation}
\widehat{f}_h(x) = \frac{1}{nh} \sum_{k=1}^n K\Big(\frac{x-x_k}{h}\Big)
(\#eq:kde)
\end{equation}

> where $K(x)$ is called the kernel function, which defines the range and size of influence of a single data point over the estimation. Most typically a simple Gaussian distribution is used as kernel function.

Reviewer: [7] specify _each_ symbol used in the equation (applicable for all following equations)

The wider the kernel function is, the farther  is the influence of a single data point, this is characterized by the bandwidth paramter $h$. 

The computational complexity of the exact KDE above is $\mathcal{O}(nm)$ where $n$ is the number of sample points to estimate from and $m$ is the number of evaluation points (the points where you want to calculate the estimate).

> There exist several methods to combat this complexity. 

Reviewer: [5] move line 17 up and combine with 15. Maybe "approximative methods to"

## Simple and linear binning

> The most straightforward way to decrease runtime is by limiting the number of sample points. This can be done by a binning routine, where the values at a smaller number of regular grid points are estimated from the original large number of sample points.

Reviewer: [3] runtime vs computational complexity is not quite the same! There is the factor parallelization ;)

Reviewer: so maybe use again computational complexity here

Given a set of sample points $X = \{x_0, x_1, ..., x_k, ..., x_{n-1}, x_n\}$ with weights $w_k$ and a set of equally spaced grid points $G = \{g_0, g_1, ..., g_l, ..., g_{n-1}, g_M\}$ where $N < n$ we can assign an estimate (or a count) $c_l$ to each grid point $g_l$ and use the newly found $g_l$'s to calculate the kernel density estimation instead. 

> This lowers the computational complexity down to $\mathcal{O}(N \cdot m)$. 

Reviewer: [4] combine with the above paragraph. It's good to have many smaller sentences, however make larger paragraphs

> Depending on the number of grid points $N$ the estimate is either more accurate and slower or less accurate and faster. However as we will see in the comparison chapter later as well, even a grid of size $1024$ is enough to capture the true density with high accuracy given a million sample points[@KDEpyDoc].

Reviewer: [3] use the word "tradeoff between" here

Simple binning is just the standard process of taking a weighted histogram that is divided by the sum of the sample points weights (normalization). In one dimension simple binning is binary in that it assigns a sample point's weight ($w_k = 1$ for an unweighted histogram) either to the grid point (bin) left or right of itself.

\begin{equation}
c_l = c(g_l) = \sum_{\substack{x_k \in X\\\frac{g_l + g_{l-1}}{2} < x_k < \frac{g_{l+1} + g_l}{2}}} w_k
(\#eq:simplebin)
\end{equation}

Linear binning on the other hand assigns a fraction of the whole weight to both grid points (bins) on either side, proportional to the closeness of grid point and data point in relation to the distance between grid points (bin width).

\begin{equation}
c_l = c(g_l) = \sum_{\substack{x_k \in X\\g_l < x_k < g_{l+1}}} \frac{g_{k+1}-x_k}{g_{l+1} - g_l} \cdot w_k + \sum_{\substack{x_k \in X\\g_{l-1} < x_k < g_l}} \frac{x_k - g_{l-1}}{g_{l+1} - g_l} \cdot w_k
(\#eq:linbin)
\end{equation}

> The kernel density estimation can then be calculated as a mixture distribution of kernel functions located at the grid points, weighted with their associated grid count.

Reviewer: grid count or bin count?

\begin{equation}
\widehat{f}_h(x) = \frac{1}{nh} \sum_{l=1}^N c_l \cdot K\Big(\frac{x-g_l}{h}\Big)
(\#eq:kdebin)
\end{equation}

## Using convolution and the Fast Fourier Transform

Another technique to speed up the computation is rewriting the Kernel Density Estimation as convolution operation between the kernel function and the grid counts calculated by the binning routine given above. 

> By using the fact that a convolution is just a multiplication in Fourier space and only evaluating the KDE at grid points one can reduce the computational complexity down to $\mathcal{O}(\log{N} \cdot N)$.[@gramacki2018fft]

Reviewer: [4] maybe something more on this? Like a motivation for it?

\begin{equation}
\widehat{f}_h(g_j) = \frac{1}{nh} \sum_{l=-N}^N k_{j-l} \cdot c_l = \vec{c} \ast \vec{k}
(\#eq:binkdeconv)
\end{equation}

where the two vectors look like this

```{r, ckFigure, fig.cap="Vectors $\\vec{c}$ and $\\vec{k}$", echo=FALSE, fig.align = 'center', out.width='60%'}
knitr::include_graphics('figures/c_conv_k.png')
```

By using the well known convolution theorem we can fourier transform $\vec{c}$ and $\vec{k}$, multiply them and inverse fourier transform them again to get the result of the discrete convolution.

However, due to the limitation of evaluating only at the grid points themselves, one needs to interpolate to get values for the estimated distribution at points in between.

## Improved Sheather-Jones Algorithm

A different take on Kernel Density Estimators is described in the paper 'Kernel density estimation by diffusion' by Botev et al.[@botev2010kernel] The authors present a new adaptive kernel density estimator based on linear diffusion processes which also includes an estimation for the optimal bandwidth.

> A more detailed and extensive explanation of the algorithm as well as an implementation in Matlab is given in the 'Handbook of Monte Carlo Methods'[@kroese2013handbook] by the original paper authors. However the general idea is briefly sketched below.

Reviewer: [3] merge with paragraph above

The optimal bandwidth is often defined as the one that minimizes the mean integrated square error($MISE$)

\begin{equation}
MISE(h) = \mathbb{E}_f\int [\widehat{f}_{h,norm}(x,h) - f(x)]^2 dx
(\#eq:mise)

> \end{equation}

Reviewer: [7] explain the symbols

To find the optimal bandwidth it is useful to look at the second order derivative $f^{(2)}$ of the unknown distribution as it indicates how many peaks the distribution has and how steep they are. For a distribution with many narrow peaks close together a smaller bandwidth leads to better result since the peaks do not get smeared together to a single peak for instance.

As derived by Wand and Jones an asymptotically optimal bandwidth $h_{AMISE}$ which minimizes a first-order asymptotic appoximation of the $MISE$ of the bandwidth is then given by[@wand1994kernel]

\begin{equation}
h_{AMISE}(x,h) = \Big( \frac{1}{2N\sqrt{\pi} \| f^{(2)}(x,h)\|^2}\Big)^{\frac{1}{5}}
(\#eq:hamise)

> \end{equation}

Reviewer: make sure all symbols are explained

(\#eq:hj)
\end{equation}

where $h_j$ is the optimal bandwidth for the $j$-th derivative of $f$.

Their proposed plug-in method works as follows:

1. Compute $\|\widehat{f}^{(l+2)}\|^2$ by assuming that f is the normal pdf with mean and variance estimated from the sample data
2. Using $\|\widehat{f}^{(l+2)}\|^2$ compute $h_{l+1}$
3. Using $h_{l+1}$ compute $\|\widehat{f}^{(l+1)}\|^2$
4. Repeat steps 2 and 3 to compute $h^{l}$, $\|\widehat{f}^{(l)}\|^2$, $h^{l-1}$, $\cdots$ and so on until 
$\|\widehat{f}^{(2)}\|^2$ is calculated
5. Use $\|\widehat{f}^{(2)}\|^2$ to compute $h_{AMISE}$

The weakest point of this procedure is the assumption that the true distribution is a Gaussian density function in order to compute $\|\widehat{f}^{(l+2)}\|^2$. This can lead to arbitrarily bad estimates of $h_{AMISE}$, when the true distribution is far from being normal.

Therefore Botev et al. took this idea further[@botev2010kernel]. Given the function $\gamma^{[k]}$ such that

\begin{equation}
\gamma^{[k]}(h)=\underbrace{\gamma_{1}\left(\cdots \gamma_{k-1}\left(\gamma_{k}\right.\right.}_{k \text { times }}(h)) \cdots)
(\#eq:gamma)

> \end{equation}

Reviewer: [6] make sure all symbols are explained

$h_{AMISE}$ can be calculated as

\begin{equation}
\begin{array}{l}
h_{AMISE} =\xi h_{1}=\xi \gamma^{[1]}\left(h_{2}\right)=\xi \gamma^{[2]}\left(h_{3}\right)=\cdots=\xi \gamma^{[l]}\left(h_{l+1}\right) \\
\xi=\left(\frac{6 \sqrt{2}-3}{7}\right)^{2 / 5} \approx 0.90
\end{array}
(\#eq:hamisegamma)
\end{equation}

By setting $h_{AMISE}=h_{l+1}$ and using fixed point iteration to solve the equation

\begin{equation}
h_{AMISE} = \xi \gamma^{[l]}(h_{AMISE})
(\#eq:hamisegamma2)
\end{equation}

the optimal bandwidth $h_{AMISE}$ can be found directly.

This eliminates the need to assume normally distributed data for the initial estimate and leads to improved performance, especially for density distributions that are far from normal as seen in the next chapter. 

> According to their paper increasing $l$ beyond $l=5$ does not increase the accuracy in any practically meaningful way.

Reviewer: Merge paragraphs

> The computation is especially efficient if $\gamma^{[5]}$ is computed using the Discrete Cosine Transform - an FFT related transformation.

Reviewer: [4] why? say something about this statements or quote

## Improved Sheather-Jones Algorithm {#isjTheory}

> A different take on kernel density estimators is described in the paper 'Kernel density estimation by diffusion' by Botev et al.[@botev2010kernel] The authors present a new adaptive kernel density estimator based on linear diffusion processes which also includes an estimation for the optimal bandwidth. A more detailed and extensive explanation of the algorithm as well as an implementation in Matlab is given in the 'Handbook of Monte Carlo Methods'[@kroese2013handbook] by the original paper authors. However the general idea is briefly sketched below.

Reviewer: ```suggestion
A different take on KDEs is described in the paper 'Kernel density estimation by diffusion' by Botev et al.[@botev2010kernel] The authors present a new adaptive kernel density estimator based on linear diffusion processes which also includes an estimation for the optimal bandwidth. A more detailed and extensive explanation of the algorithm as well as an implementation in Matlab is given in the 'Handbook of Monte Carlo Methods'[@kroese2013handbook] by the original paper authors. However the general idea is briefly sketched below.
```
